### DISCUSSION TITLE
/digging-deeper/task-scheduling.html

### DISCUSSION THREAD

**Discussion by giscus** [goravel/goravel#99] [Announcements]
# /digging-deeper/task-scheduling.html

[[toc]] Introduction In the past, you may have written a cron configuration entry for each task you needed to schedule on your server. However, this can quickly become a pain be...

https://www.goravel.dev/digging-deeper/task-scheduling.html

**Comment by narad-muni** [comment]
Why not create functions which take minutes/hours as params

**Reply by hwbrzzl** [reply]
Sorry, I missed this message. Yes, good idea, we'll have it, welcome to PR if you are interested. Please into this issue: https://github.com/goravel/goravel/issues/177

**Comment by alexisvisco** [comment]
The problem is when you scale horizontally, there is no way to use some message broker ? To ensure the job is not done two times on multiples same applications.

**Reply by hwbrzzl** [reply]
Sorry, I missed this message. Yes, you are right, we need to extend it, welcome to PR if you are interested. Please into this issue: https://github.com/goravel/goravel/issues/178

**Comment by saqijaan** [comment]
Hi @hwbrzzl, I was working on an application built with Goravel and prepared it for deployment to AWS Lambda. During the process, I encountered some issues, particularly with scheduling, the missing SQS queue driver, and a few other aspects. For now, I’d like to focus on scheduling. Currently, Goravel uses the robfig/cron package, which relies on a long-running goroutine in the background to handle schedules. However, AWS Lambda does not support this approach due to its short-lived nature. I’ve overridden the scheduler to make it compatible with Lambda. Could you review my implementation and let me know if it could be added to Goravel via a PR?

**Reply by saqijaan** [reply]
Here is the implementation. 

    // Instead of starting the cron scheduler, trigger the jobs manually
	entries := app.cron.Entries()
	for _, entry := range entries {

		// subtract 1 minute from the next time to run the job immediately
		entry.Next = entry.Schedule.Next(time.Now().Add(-1 * time.Minute))

		// If the next time is in the future, skip it
		if entry.Next.After(time.Now()) || entry.Next.IsZero() {
			facades.Log().Info("Skipping job %s, next run at %s", entry.Job, entry.Next)
			continue
		}

		entry.Job.Run() // Run the job immediately
	}

	app.log.Info("All scheduled jobs have been executed")

**Reply by saqijaan** [reply]
I've overriden the Schedular run function for now but we can add another function i.e RunNow() or RunOnce()

**Reply by hwbrzzl** [reply]
In this situation, why do you need to use the scheduling feature? What's your requirement, please?

**Reply by saqijaan** [reply]
I want to use the Goravel scheduling feature to schedule multiple jobs directly in the code. The scheduler runs every minute to check if any commands or jobs are ready to execute.

Laravel provides a way to achieve this by setting up the schedule:run command with an external system to run every minute. All other scheduling, such as daily, everyMinute, everyFiveMinutes, or everyTenMinutes, is then managed by Laravel itself.

**Reply by saqijaan** [reply]
As on Lambda we can't run any groutine forever

**Reply by hwbrzzl** [reply]
We may also provide a `schedule:run` command, User can choose to use `go facades.Schedule().Run()` or the new.

**Reply by saqijaan** [reply]
hmm but again how we will handle scheduling we will still need some sort of cron expression parser and handler as rbfig/cron does but we just don't need long running process.

**Reply by hwbrzzl** [reply]
Yes, we can also use the linux crontab to call the new `schedule:run` command, just like Laravel.

**Reply by hwbrzzl** [reply]
Created an issue: https://github.com/goravel/goravel/issues/534
